### PR TITLE
ELEMENTS-1992-Upload-Issue[LTS-2023]

### PR DESCRIPTION
**Pull request by vaibhavagarwal4-lab**
https://hyland.atlassian.net/browse/ELEMENTS-1992

### CODE REVIEW

File: ui/widgets/nuxeo-uploader-behavior.js [hunk 1]
```diff
@@ -51,49 +51,91 @@ DefaultUploadProvider.prototype._newBatch = function() {
...
+      ? Promise.all(
+          uploader._promises.map((p) =>
+            p.then(
+              () => {},
+              () => {},
+            ),
+          ),
+        )
+      : Promise.resolve();
+  return inFlight.then(() => (uploader._batchId ? uploader.cancel() : undefined));
```
[the remark this answers is not in the document]
Comment (vaibhavagarwal4-lab) [reply ui/widgets/nuxeo-uploader-behavior.js]: it is causing functionality change so ignoring this

File: ui/widgets/nuxeo-uploader-behavior.js [hunk 2]
```diff
@@ -51,49 +51,91 @@ DefaultUploadProvider.prototype._newBatch = function() {
...
+  }
+  this._ensureBatch().then(() => {
+    const uploadPromises = [];
+    for (let i = 0; i < files.length; i++) {
+      const file = files[i];
+      const blob = new Nuxeo.Blob({ content: file });
+      if (typeof callback === 'function') {
+        callback({ type: 'uploadStarted', file });
+      }
+      const p = this.uploader
```
[the remark this answers is not in the document]
Comment (vaibhavagarwal4-lab) [reply ui/widgets/nuxeo-uploader-behavior.js]: applied

File: ui/test/nuxeo-uploader-behavior.test.js [hunk 3]
```diff
@@ -199,6 +199,54 @@ suite('Nuxeo.UploaderBehavior – DefaultUploadProvider', () => {
...
+    await flushAll();
+    expect(cancel).not.to.have.been.called;
+    resolveInflight();
+    await done;
+    expect(cancel).to.have.been.calledOnce;
+  });
+
+  test('cancelBatch() still cancels when an in-flight per-blob POST rejects', async () => {
+    const rejected = Promise.reject(new Error('nope'));
+    const cancel = sinon.stub().resolves();
```
[the remark this answers is not in the document]
Comment (vaibhavagarwal4-lab) [reply ui/test/nuxeo-uploader-behavior.test.js]: applied

File: ui/widgets/nuxeo-uploader-behavior.js [hunk 4]
```diff
@@ -51,49 +51,113 @@ DefaultUploadProvider.prototype._newBatch = function() {
...
         .catch((error) => {
-          callback({ type: 'batchFailed', error, batchId: this.uploader._batchId });
+          if (!isActive()) {
+            return;
+          }
+          cb({ type: 'batchFailed', error, batchId: currentUploader._batchId });
         });
     });
-  }
+  });
```
[the remark this answers is not in the document]
Comment (vaibhavagarwal4-lab) [reply ui/widgets/nuxeo-uploader-behavior.js]: applied

File: ui/test/nuxeo-uploader-behavior.test.js [hunk 5]
```diff
@@ -231,6 +277,115 @@ suite('Nuxeo.UploaderBehavior – DefaultUploadProvider', () => {
...
+    const done = sinon.stub().resolves({ batch: { _batchId: 'b' } });
+    const uploader = { _batchId: 'b', upload, done };
+    connection.batchUpload.resolves(uploader);
+    provider.upload([fakeFile('a.txt'), fakeFile('b.txt')], cb);
+    await flushAll();
+    resolveA({ batch: { _batchId: 'b' }, blob: { fileIdx: 0 } });
+    await flushAll();
+    expect(done).not.to.have.been.called;
+    resolveB({ batch: { _batchId: 'b' }, blob: { fileIdx: 1 } });
+    await new Promise((r) => setTimeout(r, 20));
```
[the remark this answers is not in the document]
Comment (vaibhavagarwal4-lab) [reply ui/test/nuxeo-uploader-behavior.test.js]: applied

File: ui/test/nuxeo-uploader-behavior.test.js [hunk 6]
```diff
@@ -231,6 +277,115 @@ suite('Nuxeo.UploaderBehavior – DefaultUploadProvider', () => {
...
+    );
+    const done = sinon.stub().resolves({ batch: { _batchId: 'b' } });
+    const uploader = { _batchId: 'b', upload, done };
+    connection.batchUpload.resolves(uploader);
+    provider.upload([fakeFile()], cb);
+    await flushAll();
+    // Simulate cancelBatch() detaching the uploader while the POST is still pending
+    provider.uploader = null;
+    resolveUpload({ batch: { _batchId: 'b' }, blob: { fileIdx: 0 } });
+    await new Promise((r) => setTimeout(r, 20));
```
[the remark this answers is not in the document]
Comment (vaibhavagarwal4-lab) [reply ui/test/nuxeo-uploader-behavior.test.js]: added

File: ui/test/nuxeo-uploader-behavior.test.js [hunk 7]
```diff
@@ -231,6 +277,115 @@ suite('Nuxeo.UploaderBehavior – DefaultUploadProvider', () => {
...
+    };
+    connection.batchUpload.resolves(uploader);
+    provider.upload([fakeFile()], cb);
+    await flushAll();
+    // User cancels while the POST is still in flight.
+    provider.uploader = null;
+    provider.batchId = null;
+    // Server later returns 408 for the timed-out POST.
+    rejectUpload(new Error('408 Request Timeout'));
+    await new Promise((r) => setTimeout(r, 20));
```
[the remark this answers is not in the document]
Comment (vaibhavagarwal4-lab) [reply ui/test/nuxeo-uploader-behavior.test.js]: added

File: ui/test/nuxeo-uploader-behavior.test.js [hunk 8]
```diff
@@ -199,6 +199,67 @@ suite('Nuxeo.UploaderBehavior – DefaultUploadProvider', () => {
     expect(provider.uploader).to.be.null;
   });
 
+  test('cancelBatch() returns a Promise that resolves immediately (uploader.cancel() runs in the background)', async () => {
```
[the remark this answers is not in the document]
Comment (vaibhavagarwal4-lab) [reply ui/test/nuxeo-uploader-behavior.test.js]: applied.

File: ui/widgets/nuxeo-uploader-behavior.js [hunk 9]
```diff
@@ -51,49 +51,127 @@ DefaultUploadProvider.prototype._newBatch = function() {
...
+    return Promise.resolve();
   }
+  this.uploader = null;
+  this.batchId = null;
+  // Fire-and-forget: wait for any in-flight per-blob POSTs to settle, then issue the DELETE.
+  // We do NOT return this chain to the caller -- awaiting it would freeze the UI whenever a
+  // slow POST is in flight (the very thing the user just cancelled). Failures are swallowed
+  // because the batch is already detached client-side and the server will GC it on its own.
+  const inFlight =
+    uploader._promises && uploader._promises.length
```
Comment (narasimhahyland) [inline ui/widgets/nuxeo-uploader-behavior.js:163]: **This is the crux of the fix and it depends on the client's private `uploader._promises` field.** I confirmed it exists and accumulates per-blob promises (`nuxeo/lib/upload/batch.js`), so it works today. But the failure mode is silent: if the client ever renames/removes `_promises`, this ternary falls through to `Promise.resolve()`, the DELETE fires immediately again, and the ELEMENTS-1992 NPE regresses with no test catching it.

`upload()` already builds its own `uploadPromises` array — consider tracking those on the provider (e.g. `this._inFlight`) and awaiting them here instead. That fully decouples the fix from client internals and is strictly more robust. (This is the same concern Copilot raised; storing the already-existing local array on `this` is a small, safe change rather than a functional one.) At minimum, add a test asserting `_promises` is consumed so a client upgrade can't silently break the fix.

**Review by narasimhahyland — COMMENTED** [review]
Reviewed the full change (source at head, the `nuxeo` client internals it depends on, the `nuxeo-dropzone` consumer, and the tests). Overall this correctly fixes the ELEMENTS-1992 NPE: deferring `DELETE /upload/{batchId}` until every in-flight per-blob `POST` has settled removes the race that left `batch` null in `Batch.addFile(...)`, and the `isActive()` guard suppresses the spurious error toast from a cancelled batch's late 408. Tests are thorough and now deterministic via `flushAll()`. Nice work.

A few points below — one worth acting on before merge (the coupling to the client's private `_promises`), the rest are notes.

### Notes (non-blocking)
- **Cancel is not a true network abort.** Because the client has no abort, waiting for `_promises` to settle means all queued POSTs are fully uploaded and *then* the batch is DELETEd (e.g. cancel a 100-file upload at file 3 → the other 97 still upload before being discarded). This is an acceptable trade-off and no worse than before, but please add a one-line note on the ticket so QA/support know cancel doesn't stop the network transfer.
- **Edge case:** with `batchAppend: false`, if a second `upload()` starts a new batch before the first's `Promise.all` resolves, `this.uploader` changes and the first batch's `batchFinished` is suppressed by `isActive()`. Low probability and arguably fine (superseded batch), just flagging it's uncovered by tests.
- **Commit history:** please **squash-merge** — the branch has several `upload error` / `Revert` / `Reapply` commits that shouldn't land on `maintenance-3.1.x`.

File: ui/widgets/nuxeo-uploader-behavior.js (hunk 9 above)
Comment (vaibhavagarwal4-lab) [reply ui/widgets/nuxeo-uploader-behavior.js:163]: Good catch. Refactoring to this._inFlight is off the table here — we tried it and it leaked in-flight promises across batches, delaying the DELETE long enough on a slow-network cancel + retry for the server to return 408. Going with your "at minimum" suggestion instead: added a regression test that spies on [.then](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) for an entry in [_promises](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) and asserts [cancelBatch()](vscode-file://vscode-app/usr/share/code/resources/app/out/vs/code/electron-browser/workbench/workbench.html) iterates it. If a future Nuxeo client upgrade renames the field, this test fails loudly instead of the fix silently degrading. Non-blocking notes (network-abort docs, batchAppend edge case, squash-merge) noted separately.

**Review by madhurkulshrestha-hyland — APPROVED** [review]

**Review by narasimhahyland — APPROVED** [review]
